src/number_theory/fermat_psp.lean

```diff
@@ -0,0 +1,431 @@
+end
+
+/--
+All composite numbers are Fermat pseudoprimes to base 1.
+-/
+lemma pseudoprime_of_base_one (n : ℕ) (h₁ : 1 < n) (h₂ : ¬nat.prime n) : fermat_psp n 1 :=
+begin
+  refine ⟨show n ∣ 1 ^ (n - 1) - 1, from _, h₂, h₁⟩,
+  exact (show 0 = 1 ^ (n - 1) - 1, by norm_num) ▸ dvd_zero n,
+end
+
+-- For lemmas that are needed to prove statements in this file, but aren't directly related to
+-- Fermat pseudoprimes
+section helper_lemmas
+
+private lemma pow_gt_exponent (a b : ℕ) (h : 2 ≤ a) : b < a^b :=
+lt_of_lt_of_le (nat.lt_two_pow b) $ nat.pow_le_pow_of_le_left h _
+
+private lemma a_id_helper (a b : ℕ) (ha : 2 ≤ a) (hb : 2 ≤ b) : 2 ≤ (a^b - 1)/(a - 1) :=
+begin
+  change 1 < _,
+  have h₁ : a - 1 ∣ a^b - 1 := by simpa only [one_pow] using nat_sub_dvd_pow_sub_pow a 1 b,
+  rw [nat.lt_div_iff_mul_lt h₁, mul_one, tsub_lt_tsub_iff_right (nat.le_of_succ_le ha)],
+  convert pow_lt_pow (nat.lt_of_succ_le ha) hb,
+  rw pow_one
+end
+
+private lemma b_id_helper (a b : ℕ) (ha : 2 ≤ a) (hb : 2 < b) : 2 ≤ (a^b + 1)/(a + 1) :=
+begin
+  rw nat.le_div_iff_mul_le (nat.zero_lt_succ _),
+  apply nat.succ_le_succ,
+  calc 2 * a + 1 ≤ a ^ 2 * a : by nlinarith
+             ... = a ^ 3     : by rw pow_succ' a 2
+             ... ≤ a ^ b     : pow_le_pow (nat.le_of_succ_le ha) hb
+end
+
+private lemma AB_id_helper (b p : ℕ) (hb : 2 ≤ b) (hp : odd p)
+  : ((b ^ p - 1) / (b - 1)) * ((b ^ p + 1) / (b + 1)) = ((b ^ (2*p)) - 1) / (b^2 - 1) :=
+begin
+  have q₁ : b - 1 ∣ b ^ p - 1 := by simpa only [one_pow] using nat_sub_dvd_pow_sub_pow b 1 p,
+  have q₂ : b + 1 ∣ b ^ p + 1 := by simpa only [one_pow] using hp.nat_add_dvd_pow_add_pow b 1,
+  convert nat.div_mul_div_comm q₁ q₂; rw [mul_comm (_ - 1), ←nat.sq_sub_sq],
+  { ring_exp },
+  { simp }
+end
+
+end helper_lemmas
+
+/--
+Given a prime `p` which does not divide `b*(b^2 - 1)`, we can produce a number `n` which is larger
+than `p` and pseudoprime to base `b`. We do this by defining
+`n = ((b^p - 1)/(b - 1)) * ((b^p + 1)/(b + 1))`
+
+The primary purpose of this definition is to help prove `exists_infinite_pseudoprimes`. For a proof
+that `n` is actually pseudoprime to base `b`, see `psp_from_prime_psp`, and for a proof that `n` is
+greater than `p`, see `psp_from_prime_gt_p`.
+
+This lemma is intended to be used when `2 ≤ b`, `2 < p`, `p` is prime, and `¬p ∣ b*(b^2 - 1)`,
+because those are the hypotheses for `psp_from_prime_psp`.
+-/
+private def psp_from_prime (b : ℕ) (p : ℕ) : ℕ := (b^p - 1)/(b - 1) * ((b^p + 1)/(b + 1))
+
+/--
+This is a proof that the number produced using `psp_from_prime` is actually pseudoprime to base `b`.
+The primary purpose of this lemma is to help prove `exists_infinite_pseudoprimes`.
+
+We use <https://primes.utm.edu/notes/proofs/a_pseudoprimes.html> as a rough outline of the proof.
+-/
+private lemma psp_from_prime_psp (b : ℕ) (b_ge_two : 2 ≤ b) (p : ℕ) (p_prime : nat.prime p)
+  (p_gt_two : 2 < p) (not_dvd : ¬p ∣ b*(b^2 - 1)) :
+  fermat_psp (psp_from_prime b p) b :=
+begin
```

Comment:
This proof is quite long, do you think it is possible to extract some sublemmas?